### ISSUE TITLE
Support for java.util.ImmutableCollections

### ISSUE THREAD

**Issue by efwe** [EsotericSoftware/kryo#693]
With Java9 we got new factory methods like `Set.of`,`List.of` and `Map.of` which use the (non public) class `java.util.ImmutableCollections`. Without a special serializer I get the following error:
```
com.esotericsoftware.kryo.KryoException: java.lang.UnsupportedOperationException
Serialization trace:
longList (de.fw.KryoTest$Immutables)

	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:144)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:543)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:709)
	at de.fw.KryoTest.testSerializeJava9ImmutableCollections(KryoTest.java:54)
Caused by: java.lang.UnsupportedOperationException
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:134)
	at com.esotericsoftware.kryo.serializers.CollectionSerializer.read(CollectionSerializer.java:40)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:731)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
	... 3 more
```

When I configure Kryo like this 
`kryo.register(List.of().getClass(), new JavaSerializer());`
it works. 

My questions are: 
* Do any special serializers already exist?
* Should they exist? 

So this is half question, half bug report here.
Keep up the good work,
Frank

**Comment by theigl** [comment]
I just ran into this issue as well. There should definitely be serializers for these new immutable collections. They can probably be implemented like the immutable collection serializers for Guava in https://github.com/magro/kryo-serializers.

See for instance https://github.com/magro/kryo-serializers/blob/master/src/main/java/de/javakaffee/kryoserializers/guava/ImmutableListSerializer.java

**Comment by theigl** [comment]
Here is a quick implementation of serializers for `java.util.ImmutableCollections`: https://gist.github.com/theigl/7979208a36170b40060ec0c9c4690104

**Comment by theigl** [comment]
@magro: Are you open to creating a JDK11 version of kryo-serializers? If so, I could create a PR with my serializers for `java.util.ImmutableCollections`.

**Comment by magro** [comment]
@theigl can you please submit an issue for kryo-serializers for this? Then I hope to have a look at this next week...

**Comment by magro** [comment]
Hrm, thinking again, because this is about jdk stuff and not 3rd party libs it should rather be added to kryo instead of kryo-serializers.

So we could just keep it here.

**Comment by theigl** [comment]
@magro: I agree that this should eventually be part of Kryo itself. The issue is that Kryo currently has a JDK8 baseline and the serializers need the new collection methods on the classpath. It would theoretically be possible to register them using reflection, but it would make things much more complicated.

There are two options I currently see for adding them directly to Kryo:

1. Start a new Kryo 6.x branch with JDK11 as a baseline and add the serializers there
2. Build an additional artifact with JDK11 support for Kryo 5

Alternatively, we could do either 1. or 2. for `kryo-serializers` where you have full control, add the serializers there and move them to Kryo later on.

In the long run it makes a lot of sense to move Kryo to JDK11 and test against newer non-LTS releases as well. 

I compiled and tested Kryo against JDK11 and JDK14 yesterday. There are only two test failures and Unsafe ByteBuffers do not work. I can start working on a branch with full JDK11 compatibility next week.

**Comment by magro** [comment]
My idea was to follow the approach at we used for java8 serializers when kryo was still supporting java 6: https://github.com/EsotericSoftware/kryo/pull/362

WDYT? 

Would be cool if you could check in parallel jdk11 compatibility (so I could do just the build / maven stuff).

**Comment by theigl** [comment]
That sounds good! So we switch to JDK11 for the build but target JDK8 for the release. In that case, I still have to fix the tests currently failing with JDK11.

I'll create two separate PRs next week. One fixing the tests when building with JDK11 and one adding the new serializers with the method you suggested. It would be great if you could take care of the build.

**Comment by theigl** [comment]
@magro: I just pushed the PR to fix tests on JDK 11 #719.

**Comment by magro** [comment]
@theigl Cool, let's do it as you said.

**Comment by theigl** [comment]
@magro: Unfortunately, we cannot take the same approach as for Java 8. JDK9+ introduced incompatible changes for `ByteBuffer` methods that result in errors when tests are run with JDK8:

```
java.lang.NoSuchMethodError: java.nio.ByteBuffer.position(I)Ljava/nio/ByteBuffer;
	at com.esotericsoftware.kryo.DepthTest.testDepth(DepthTest.java:42)
```

See also https://github.com/eclipse/jetty.project/issues/3244

It is possible to work around this with explicit casts but this will pollute the code quite a bit.

The correct way to resolve this is the `--release` option when compiling with newer versions of the JDK, but if we use it, the new `java.util.ImmutableCollections` classes are not available for the build.

I suspect that we have to use one of the options [outlined above](https://github.com/EsotericSoftware/kryo/issues/693#issuecomment-629767200) if we want to resolve this.

**Comment by theigl** [comment]
I have prepared a [branch](https://github.com/EsotericSoftware/kryo/compare/master...theigl:jdk11-immutable-collections?expand=1) if you want to take a look at it yourself.

**Comment by magro** [comment]
Hey @theigl, that's really annoying! I checked out your branch, and tried to get it to work with different executions in the maven-compiler-plugin, but couldn't solve it with this so far. When I have more time I might continue to check this out, but I cannot promise.

The options outlined above (additional kryo 6 branch or another artifact) don't sound very attractive to me... Not sure...

**Comment by theigl** [comment]
> Hey @theigl, that's really annoying! 

I know. I wasted two hours on this yesterday ;)

> The options outlined above (additional kryo 6 branch or another artifact) don't sound very attractive to me... Not sure...

Same here. Deciding on a JDK11 baseline for Kryo 6 probably makes sense, but that should wait until the final release of version 5.

I'll investigate some more later today. Maybe it is possible to wrap all ByteBuffer usages in a `BackwardsCompatibleByteBuffer` that abstracts away the casts required to run on JDK8.

**Comment by magro** [comment]
> Maybe it is possible to wrap all ByteBuffer usages in a `BackwardsCompatibleByteBuffer` that abstracts away the casts required to run on JDK8.

Right, or some ByteBufferUtil that encapsulates this. I'd expect that this would be inlined by JIT and not really creates a performance overhead.

**Comment by theigl** [comment]
I pushed an experimental change https://github.com/theigl/kryo/commit/4057a6d1f98895107d547079e841da7c99bb0150 that allows us to run the tests with JDK8. It doesn't look too great, but it would get the job done.

**Comment by magro** [comment]
That's huge! Do you want to submit this as a PR?

So far I only see a difference/inconsistency between java 9 and java 11 in some places, where I think we could refer to java 9 everywhere (e.g. also in this testdata stuff). But we could discuss this as part of the PR.

**Comment by theigl** [comment]
@magro: Sure! I think I'll split it into two separate PRs. The first one will be about the JDK11 build/test infrastructure only. So we can discuss this, before I continue working on the serializers.

**Comment by theigl** [comment]
@magro: I just pushed #724

**Comment by theigl** [comment]
#727 with support for immutable collections is ready for review.
